File: app/src/main/java/com/pandatone/kumiwake/member/FragmentGroup.java
package com.pandatone.kumiwake.member;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.ListFragment;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;
import android.util.Log;
import android.util.SparseBooleanArray;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;

import com.pandatone.kumiwake.R;
import com.pandatone.kumiwake.adapter.GroupNameListAdapter;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by atsushi_2 on 2016/02/23.
 */
public class FragmentGroup extends ListFragment {
    private MemberMain parent;
    static GroupListAdapter dbAdapter;
    static GroupNameListAdapter listAdapter;
    static List<GroupListAdapter.Group> nameList;
    static GroupListAdapter.Group listItem;
    static FloatingActionButton fab;
    static TextView adviceInFG;
    static int ListCount, checkedCount=0;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        dbAdapter = new GroupListAdapter(getActivity());
        nameList = new ArrayList<>();
        listAdapter = new GroupNameListAdapter(getActivity(), nameList);
        setListAdapter(listAdapter);
        loadName();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.tab_group, container, false);
        adviceInFG = (TextView) view.findViewById(R.id.advice_in_fg);
        fab = (FloatingActionButton) view.findViewById(R.id.group_fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                parent.moveGroup();
            }
        });
        return view;
    }


    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        ListCount = getListView().getCount();
        getListView().setChoiceMode(ListView.CHOICE_MODE_MULTIPLE_MODAL);
        getListView().setMultiChoiceModeListener(new Callback());
        getListView().setFastScrollEnabled(true);
        ListCount = getListView().getCount();

        getListView().setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            //行をクリックした時の処理
            public void onItemClick(final AdapterView<?> parent, View view, final int position, long id) {

                final android.support.v7.app.AlertDialog.Builder builder = new android.support.v7.app.AlertDialog.Builder(getActivity());
                final android.app.AlertDialog.Builder builder2 = new android.app.AlertDialog.Builder(getActivity());
                LayoutInflater inflater = (LayoutInflater) getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
                final View view2 = inflater.inflate(R.layout.group_info,
                        (ViewGroup) getActivity().findViewById(R.id.info_layout));
                final String groupname = nameList.get(position).getGroup();
                if (MemberMain.searchView.isActivated() == true)
                    MemberMain.searchView.onActionViewCollapsed();
                FragmentMember.loadName();

                final String[] Items =
                        {Sort.name_getContext().getString(R.string.information),
                                Sort.name_getContext().getString(R.string.edit),
                                Sort.name_getContext().getString(R.string.delete),};
                builder.setTitle(groupname);
                builder.setItems(Items, new DialogInterface.OnClickListener() {
                    public void onClick(final DialogInterface dialog, int which) {
                        switch (which) {
                            case 0:
                                GroupClick.GroupInfoDialog(view2, builder2);
                                GroupClick.SetInfo(position);
                                final android.app.AlertDialog dialog2 = builder2.create();
                                dialog2.show();
                                GroupClick.okBt.setOnClickListener(new View.OnClickListener() {
                                    public void onClick(View v) {
                                        dialog2.dismiss();
                                    }
                                });
                                break;
                            case 1:
                                Intent i = new Intent(getActivity(), AddGroup.class);
                                i.putExtra("POSITION", position);
                                startActivity(i);
                                break;
                            case 2:
                                DeleteSingleGroup(position, groupname);
                                break;

                        }
                    }
                });
                android.support.v7.app.AlertDialog dialog = builder.create();
                dialog.show();
            }
        });

        if (parent.start_actionmode == true) {
            getListView().setOnItemClickListener(new AdapterView.OnItemClickListener() {
                @Override
                //行をクリックした時の処理
                public void onItemClick(final AdapterView<?> parent, View view, final int position, long id) {
                    getListView().startActionMode(new Callback());
                    getListView().setItemChecked(position, !listAdapter.isPositionChecked(position));
                }
            });
        }

        getListView().setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {

            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, final int position, long id) {
                getListView().setItemChecked(position, !listAdapter.isPositionChecked(position));
                return false;
            }
        });

        getListView().setTextFilterEnabled(true);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // アクションアイテム選択時
        switch (item.getItemId()) {
            case android.R.id.home:
                getActivity().finish();
                break;

            case R.id.item_delete:
                DeleteGroup();
                break;

            case R.id.item_all_select:
                for (int i = 0; i < ListCount; i++) {
                    getListView().setItemChecked(i, true);
                }
                break;

            case R.id.item_sort:
                final android.support.v7.app.AlertDialog.Builder builder = new android.support.v7.app.AlertDialog.Builder(getActivity());
                Sort.groupSort(builder);
                android.support.v7.app.AlertDialog dialog = builder.create();
                dialog.show();
                ListCount = getListView().getCount();
                break;

        }

        return false;
    }

    @Override
    public void onAttach(Context context) {
        parent = (MemberMain) context;
        super.onAttach(context);
    }

    @Override
    public void onStart() {
        super.onStart();
        loadName();
    }


    public static void loadName() {
        nameList.clear();
        dbAdapter.open();
        Cursor c = dbAdapter.getAllNames();
        dbAdapter.getCursor(c);
        dbAdapter.close();
    }

    public static void selectGroup(String newText) throws IOException {
        if (TextUtils.isEmpty(newText)) {
            dbAdapter.picGroup(null,null);
        } else {
            dbAdapter.picGroup(newText,newText);
        }

    }

    public void DeleteSingleGroup(final int position, String group) {
        final android.support.v7.app.AlertDialog.Builder builder = new android.support.v7.app.AlertDialog.Builder(getActivity());
        builder.setTitle(group);
        builder.setMessage(R.string.Do_delete);
        // OKの時の処理
        builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dbAdapter.open();
                listItem = nameList.get(position);
                int listId = listItem.getId();
                dbAdapter.selectDelete(String.valueOf(listId));
                dbAdapter.close();    // DBを閉じる
                loadName();
            }
        });

        builder.setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
            }
        });
        android.support.v7.app.AlertDialog dialog = builder.create();
        dialog.show();
    }

    public void DeleteGroup() {
        Log.d("DeleteGroup is called.", "hooo");
        // アラートダイアログ表示
        final AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(checkedCount + " " + getString(R.string.group) + getString(R.string.delete));
        builder.setMessage(R.string.Do_delete);
        // OKの時の処理
        builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                final SparseBooleanArray list = getListView().getCheckedItemPositions();

                dbAdapter.open();     // DBの読み込み(読み書きの方)
                for (int i = 0; i < ListCount; i++) {
                    boolean checked = list.get(i);
                    if (checked == true) {
                        // IDを取得する
                        listItem = nameList.get(i);
                        int listId = listItem.getId();
                        dbAdapter.selectDelete(String.valueOf(listId));     // DBから取得したIDが入っているデータを削除する
                        FragmentMember.DeleteBelongInfoAll(listId);
                    }
                }
                dbAdapter.close();    // DBを閉じる
                listAdapter.clearSelection();
                getListView().setChoiceMode(ListView.CHOICE_MODE_NONE);
                getListView().setChoiceMode(ListView.CHOICE_MODE_MULTIPLE_MODAL);
                loadName();
            }
        });

        builder.setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();

        getListView().setTextFilterEnabled(true);
    }

    //////////////////////////////////////////////////////////////////////////////////////
    /////////////////////-------- ActionMode時の処理 ----------///////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////


    private class Callback implements ListView.MultiChoiceModeListener {

        final SparseBooleanArray list = getListView().getCheckedItemPositions();

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            // アクションモード初期化処理
            MenuInflater inflater = getActivity().getMenuInflater();
            inflater.inflate(R.menu.member_main_menu, menu);
            menu.getItem(2).setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
            menu.getItem(3).setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
            menu.getItem(4).setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
            parent.decision.setOnClickListener(decision_clicked);
            MenuItem searchIcon = menu.findItem(R.id.search_view);
            MenuItem deleteIcon = menu.findItem(R.id.item_delete);
            MenuItem itemfilter = menu.findItem(R.id.item_filter);
            itemfilter.setVisible(false);
            searchIcon.setVisible(false);
            deleteIcon.setVisible(parent.delete_icon_visible);
            return true;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            // アクションアイテム選択時
            switch (item.getItemId()) {
                case R.id.item_delete:
                    DeleteGroup();
                    break;

                case R.id.item_all_select:
                    for (int i = 0; i < ListCount; i++) {
                        getListView().setItemChecked(i, true);
                    }
                    break;

                case R.id.item_sort:
                    final android.support.v7.app.AlertDialog.Builder builder = new android.support.v7.app.AlertDialog.Builder(getActivity());
                    Sort.groupSort(builder);
                    android.support.v7.app.AlertDialog dialog = builder.create();
                    dialog.show();
                    ListCount = getListView().getCount();
                    break;
            }
            return false;
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            // 決定ボタン押下時
            listAdapter.clearSelection();
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            // アクションモード表示事前処理
            return true;
        }

        @Override
        public void onItemCheckedStateChanged(ActionMode mode,
                                              int position, long id, boolean checked) {
            // アクションモード時のアイテムの選択状態変更時

            checkedCount=getListView().getCheckedItemCount();

            if (checked) {
                listAdapter.setNewSelection(position, checked);
            } else {
                listAdapter.removeSelection(position);
            }
            mode.setTitle(checkedCount + getString(R.string.selected));
        }

        private View.OnClickListener decision_clicked = new View.OnClickListener() {

            public void onClick(View v) {

                dbAdapter.open();     // DBの読み込み(読み書きの方)
                if (parent.kumiwake_select == false) {
                    for (int i = 0; i < ListCount; i++) {
                        boolean checked = list.get(i);
                        if (checked == true) {
                            listItem = nameList.get(i);
                            int myId = listItem.getId();

                            int newId = parent.groupId;
                            FragmentMember.addGroupByGroup(newId, myId);
                        }
                    }
                    parent.finish();
                } else {

                    for (int i = 0; i < ListCount; i++) {
                        boolean checked = list.get(i);
                        if (checked == true) {
                            listItem = nameList.get(i);
                            int myId = listItem.getId();
                            FragmentMember.createKumiwakeListByGroup(myId);
                        }
                    }
                    parent.moveKumiwake();
                }
                listAdapter.clearSelection();
                dbAdapter.close();
            }
        };

    }

}

File: app/src/main/java/com/pandatone/kumiwake/member/FragmentGroup.kt
package com.pandatone.kumiwake.member

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.*
import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.ListFragment
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.pandatone.kumiwake.MyApplication
import com.pandatone.kumiwake.R
import com.pandatone.kumiwake.adapter.GroupListAdapter
import com.pandatone.kumiwake.adapter.GroupNameListAdapter
import java.io.IOException

/**
 * Created by atsushi_2 on 2016/02/23.
 */
class FragmentGroup : ListFragment() {
    private lateinit var listItem: GroupListAdapter.Group
    private var checkedCount = 0

    // 必須*
    // Fragment生成時にシステムが呼び出す
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        dbAdapter = GroupListAdapter(requireContext())
        groupList = ArrayList()
        listAdp = GroupNameListAdapter(requireContext(), groupList)
    }

    // 必須*
    // Fragmentが初めてUIを描画する時にシステムが呼び出す
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.tab_group, container, false)
        adviceInFG = view.findViewById<View>(R.id.advice_in_fg) as TextView
        fab = view.findViewById<View>(R.id.group_fab) as FloatingActionButton
        fab.setOnClickListener { moveGroup() }

        // Fragmentとlayoutを紐付ける
        super.onCreateView(inflater, container, savedInstanceState)
        return view
    }

    private fun moveGroup() {
        val intent = Intent(activity, AddGroup::class.java)
        startActivity(intent)
    }

    // Viewの生成が完了した後に呼ばれる
    // UIパーツの設定などを行う
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        listView.choiceMode = ListView.CHOICE_MODE_MULTIPLE_MODAL
        listView.setMultiChoiceModeListener(Callback())
        listView.isFastScrollEnabled = true

        listView.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            //行をクリックした時の処理
            val builder = androidx.appcompat.app.AlertDialog.Builder(activity!!)
            val builder2 = android.app.AlertDialog.Builder(activity)
            val inflater = activity!!.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val view2 = inflater.inflate(R.layout.group_info, activity!!.findViewById<View>(R.id.info_layout) as ViewGroup?)
            val groupName = groupList[position].group
            if (MemberMain.searchView.isActivated)
                MemberMain.searchView.onActionViewCollapsed()
            FragmentMember().loadName()

            val items = arrayOf(MyApplication.context?.getString(R.string.information), MyApplication.context?.getString(R.string.edit), MyApplication.context?.getString(R.string.delete))
            builder.setTitle(groupName)
            builder.setItems(items) { _, which ->
                when (which) {
                    0 -> {
                        GroupClick.groupInfoDialog(view2, builder2)
                        GroupClick.setInfo(position)
                        val dialog2 = builder2.create()
                        dialog2.show()
                        GroupClick.okBt.setOnClickListener { dialog2.dismiss() }
                    }
                    1 -> {
                        val i = Intent(activity, AddGroup::class.java)
                        i.putExtra(AddGroup.GROUP_ID, groupList[position].id)
                        startActivity(i)
                    }
                    2 -> deleteSingleGroup(position, groupName)
                }
            }
            val dialog = builder.create()
            dialog.show()
        }

        listView.onItemLongClickListener = AdapterView.OnItemLongClickListener { _, _, position, _ ->
            listView.setItemChecked(position, !listAdp.isPositionChecked(groupList[position].id))
            false
        }

        listView.isTextFilterEnabled = true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        // アクションアイテム選択時
        when (item!!.itemId) {
            android.R.id.home -> activity!!.finish()

            R.id.item_all_select -> for (i in 0 until listAdp.count) {
                listView.setItemChecked(i, true)
            }

            R.id.item_sort -> {
                val builder = androidx.appcompat.app.AlertDialog.Builder(activity!!)
                Sort.groupSort(builder, activity!!)
                val dialog = builder.create()
                dialog.show()
            }
        }

        return false
    }

    //Activity生成後に呼ばれる
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (MemberMain.startAction) {
            listView.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
                //行をクリックした時の処理
                MemberMain.viewPager.setCurrentItem(0, true)
                FragmentMember().checkByGroup(groupList[position].id)
            }
        }
    }

    override fun onStart() {
        super.onStart()
        loadName()
        FragmentMember().loadName()
    }

    private fun deleteSingleGroup(position: Int, group: String) {
        val builder = androidx.appcompat.app.AlertDialog.Builder(activity!!)
        builder.setTitle(group)
        builder.setMessage(R.string.Do_delete)
        // OKの時の処理
        builder.setPositiveButton("OK") { _, _ ->
            dbAdapter.open()
            listItem = groupList[position]
            val listId = listItem.id
            dbAdapter.selectDelete(listId.toString())
            dbAdapter.close()    // DBを閉じる
        }

        builder.setNegativeButton(R.string.cancel) { _, _ -> }
        val dialog = builder.create()
        dialog.show()

        loadName()
    }

    fun deleteGroup(mode: ActionMode) {

        // アラートダイアログ表示
        val builder = AlertDialog.Builder(activity!!)
        builder.setTitle(checkedCount.toString() + " " + getString(R.string.group) + getString(R.string.delete))
        builder.setMessage(R.string.Do_delete)
        // OKの時の処理
        builder.setPositiveButton("OK") { _, _ ->
            val list = listView.checkedItemPositions

            dbAdapter.open()     // DBの読み込み(読み書きの方)
            for (i in 0 until listAdp.count) {
                val checked = list.get(i)
                if (checked) {
                    // IDを取得する
                    listItem = groupList[i]
                    val listId = listItem.id
                    FragmentMember().deleteBelongInfoAll(listId)
                    dbAdapter.selectDelete(listId.toString())     // DBから取得したIDが入っているデータを削除する
                }
            }
            dbAdapter.close()    // DBを閉じる
            listAdp.clearSelection()
            FragmentMember().loadName()
            loadName()
            mode.finish()
        }

        builder.setNegativeButton(R.string.cancel) { _, _ -> }
        val dialog = builder.create()
        dialog.show()

        listView.isTextFilterEnabled = true
    }

    //////////////////////////////////////////////////////////////////////////////////////
    /////////////////////-------- ActionMode時の処理 ----------///////////////////////////
    //////////////////////////////////////////////////////////////////////////////////////


    private inner class Callback : AbsListView.MultiChoiceModeListener {

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            // アクションモード初期化処理
            val inflater = activity!!.menuInflater
            inflater.inflate(R.menu.member_main_menu, menu)
            menu.getItem(2).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
            menu.getItem(3).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
            menu.getItem(4).setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
            val searchIcon = menu.findItem(R.id.search_view)
            val deleteIcon = menu.findItem(R.id.item_delete)
            val itemFilter = menu.findItem(R.id.item_filter)
            itemFilter.isVisible = false
            searchIcon.isVisible = false
            deleteIcon.isVisible = MemberMain.delete_icon_visible
            return true
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            // アクションアイテム選択時
            when (item.itemId) {
                R.id.item_delete -> {
                    deleteGroup(mode)
                }

                R.id.item_all_select -> for (i in 0 until listAdp.count) {
                    listView.setItemChecked(i, false)
                }

                R.id.item_sort -> {
                    listView.clearChoices()
                    listAdp.clearSelection()
                    mode.title = "0" + getString(R.string.selected)
                    val builder = androidx.appcompat.app.AlertDialog.Builder(activity!!)
                    Sort.groupSort(builder, activity!!)
                    val dialog = builder.create()
                    dialog.show()
                }
            }
            return false
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            // 決定ボタン押下時
            listAdp.clearSelection()
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
            // アクションモード表示事前処理
            return true
        }

        override fun onItemCheckedStateChanged(mode: ActionMode,
                                               position: Int, id: Long, checked: Boolean) {
            // アクションモード時のアイテムの選択状態変更時

            checkedCount = listView.checkedItemCount
            listView.checkedItemPosition

            if (checked) {
                listAdp.setNewSelection(groupList[position].id, checked)
            } else {
                listAdp.removeSelection(groupList[position].id)
            }
            mode.title = checkedCount.toString() + getString(R.string.selected)
        }

    }

    @Throws(IOException::class)
    fun selectGroup(newText: String) {
        if (TextUtils.isEmpty(newText)) {
            dbAdapter.picGroup(null.toString(), null.toString())
        } else {
            dbAdapter.picGroup(newText, newText)
        }

    }

    fun loadName() {
        dbAdapter.open()
        val c = dbAdapter.getDB
        dbAdapter.getCursor(c, groupList)
        dbAdapter.close()
        listAdapter = listAdp
        listAdp.notifyDataSetChanged()
    }

    companion object {
        internal lateinit var listAdp: GroupNameListAdapter
        internal lateinit var dbAdapter: GroupListAdapter
        internal var groupList: ArrayList<GroupListAdapter.Group> = ArrayList()
        internal lateinit var fab: FloatingActionButton
        internal lateinit var adviceInFG: TextView
    }

}






